pkg/codegen/internal/test/testdata/env-helper/nodejs/types/input.ts

```diff
@@ -0,0 +1,97 @@
+// *** WARNING: this file was generated by test. ***
+// *** Do not edit by hand unless you're certain you know what you are doing! ***
+
+import * as pulumi from "@pulumi/pulumi";
+import { input as inputs, output as outputs } from "../types";
+
+import * as utilities from "../utilities";
+
+/**
+ * BETA FEATURE - Options to configure the Helm Release resource.
+ */
+export interface HelmReleaseSettingsArgs {
+    /**
+     * The backend storage driver for Helm. Values are: configmap, secret, memory, sql.
+     */
+    driver?: pulumi.Input<string>;
+    /**
+     * The path to the helm plugins directory.
+     */
+    pluginsPath?: pulumi.Input<string>;
+    /**
+     * to test required args
+     */
+    requiredArg: pulumi.Input<string>;
+}
+/**
+ * helmReleaseSettingsArgsProvideDefaults sets the appropriate defaults for HelmReleaseSettingsArgs
+ */
+export function helmReleaseSettingsArgsProvideDefaults(val: HelmReleaseSettingsArgs): HelmReleaseSettingsArgs {
+    return {
+        ...val,
+    driver: (val.driver) ?? (utilities.getEnv("PULUMI_K8S_HELM_DRIVER") || "secret"),
+    pluginsPath: (val.pluginsPath) ?? utilities.getEnv("PULUMI_K8S_HELM_PLUGINS_PATH"),
+    };
+}
+
+/**
+ * Options for tuning the Kubernetes client used by a Provider.
+ */
+export interface KubeClientSettingsArgs {
+    /**
+     * Maximum burst for throttle. Default value is 10.
+     */
+    burst?: pulumi.Input<number>;
+    /**
+     * Maximum queries per second (QPS) to the API server from this client. Default value is 5.
+     */
+    qps?: pulumi.Input<number>;
+}
+/**
+ * kubeClientSettingsArgsProvideDefaults sets the appropriate defaults for KubeClientSettingsArgs
+ */
+export function kubeClientSettingsArgsProvideDefaults(val: KubeClientSettingsArgs): KubeClientSettingsArgs {
+    return {
+        ...val,
+    burst: (val.burst) ?? utilities.getEnvNumber("PULUMI_K8S_CLIENT_BURST"),
+    qps: (val.qps) ?? utilities.getEnvNumber("PULUMI_K8S_CLIENT_QPS"),
+    };
+}
+
+/**
+ * Make sure that defaults propagate through types
+ */
+export interface LayeredTypeArgs {
+    /**
+     * The answer to the question
+     */
+    answer?: pulumi.Input<number>;
+    other: pulumi.Input<inputs.HelmReleaseSettingsArgs>;
+    /**
+     * Test how plain types interact
+     */
+    plainOther?: inputs.HelmReleaseSettingsArgs;
+    /**
+     * The question already answered
+     */
+    question?: pulumi.Input<string>;
+    recursive?: pulumi.Input<inputs.LayeredTypeArgs>;
+    /**
+     * To ask and answer
+     */
+    thinker: pulumi.Input<string>;
+}
+/**
+ * layeredTypeArgsProvideDefaults sets the appropriate defaults for LayeredTypeArgs
+ */
+export function layeredTypeArgsProvideDefaults(val: LayeredTypeArgs): LayeredTypeArgs {
+    return {
+        ...val,
+    answer: (val.answer) ?? 42,
+    other: pulumi.output(val.other).apply(inputs.helmReleaseSettingsArgsProvideDefaults)!,
```

Comment:
What does `!` do is this ignoring a potential null?

Reply:
It would  never be null (only applied to inputs that are not nullable). The typescript type systems apparently knows this. I experimented with removing the `!` and tests passed. I think I can remove the `!`.